File: packages/adapters/src/queue.ts
/**
 * Queue contract — the producer side of an asynchronous job/message queue.
 *
 * Mirrors `fiducial_adapters::queue` in Rust for the contract and `none`.
 * `cloudflare-queues` is real only here: a Cloudflare Queue producer binding
 * exists only inside a Worker. See the Rust module's doc comment for why
 * there is no `receive`/`consume` method on this contract at all — a
 * consumer is an exported handler, not a call an adapter makes.
 */

export class QueueError extends Error {
  constructor(message: string) {
    super(message);
    this.name = "QueueError";
  }
}

/** Asynchronous queue contract — producer side only. */
export interface Queue {
  /** Enqueue a single message body. */
  send(body: Uint8Array): Promise<void>;

  /**
   * Enqueue several message bodies as one batch.
   *
   * No atomicity promise across the batch — unlike `Database.batch`, no
   * vendor behind this contract has a real batch/transaction primitive to
   * lean on for it.
   */
  sendBatch(bodies: Uint8Array[]): Promise<void>;
}

// ── None implementation ───────────────────────────────────────────────────────

/** No-op queue — every send succeeds silently and the message is discarded. */
export class NoneQueue implements Queue {
  // Accepts and ignores `env`, matching every other vendor class in this
  // contract, for the generated factory to call uniformly.
  constructor(_env?: unknown) {}

  async send(_body: Uint8Array): Promise<void> {}

  async sendBatch(_bodies: Uint8Array[]): Promise<void> {}
}

// ── Cloudflare Queues ────────────────────────────────────────────────────────

/**
 * `CloudflareBinding` mirrors the methods this adapter calls from
 * `@cloudflare/workers-types`' `Queue`, without depending on that package at
 * the type level — same reasoning as `D1Binding`/`R2Binding` in the other
 * two contracts.
 */
interface CloudflareQueueBinding {
  send(message: unknown, options?: { contentType?: string }): Promise<void>;
  sendBatch(
    messages: Iterable<{ body: unknown; contentType?: string }>,
  ): Promise<void>;
}

/**
 * Cloudflare Queues — reached through a binding (`env.QUEUE`) inside a
 * Worker. Bytes are sent with `contentType: "bytes"` so the consumer handler
 * receives the same `Uint8Array` back rather than Cloudflare's default JSON
 * round-trip.
 *
 * **Binding convention:** `env.QUEUE`, for the same reason `D1Database`
 * reads `env.DB` — `fid derive` cannot see a hand-edited `wrangler.toml`.
 */
export class CloudflareQueue implements Queue {
  private readonly queue: CloudflareQueueBinding;

  constructor(env: { QUEUE?: CloudflareQueueBinding }) {
    if (!env?.QUEUE) {
      throw new QueueError(
        "CloudflareQueue: no `QUEUE` binding on env — add a [[queues.producers]] " +
          'block with `binding = "QUEUE"` to wrangler.toml',
      );
    }
    this.queue = env.QUEUE;
  }

  async send(body: Uint8Array): Promise<void> {
    try {
      await this.queue.send(body, { contentType: "bytes" });
    } catch (err) {
      throw new QueueError(`Cloudflare Queues send failed: ${String(err)}`);
    }
  }

  async sendBatch(bodies: Uint8Array[]): Promise<void> {
    try {
      await this.queue.sendBatch(
        bodies.map((body) => ({ body, contentType: "bytes" })),
      );
    } catch (err) {
      throw new QueueError(
        `Cloudflare Queues sendBatch failed: ${String(err)}`,
      );
    }
  }
}

File: crates/fiducial-adapters/src/queue.rs
//! Queue contract — the producer side of an asynchronous job/message queue.
//!
//! Designed against the narrowest interface shared by Cloudflare Queues, SQS
//! and a Postgres-backed queue: "hand me a message, it gets delivered." A
//! consumer is deliberately **not** part of this contract — see below.
//!
//! **Only `none` ships a Rust implementation**, and for a structural reason
//! this time, not a "nobody has asked yet" one: `cloudflare-queues` — the one
//! real vendor — is reached through a Workers binding
//! (`env.QUEUE.send()`/`sendBatch()`), which exists only inside a Worker, the
//! same boundary `d1` and `r2` sit behind. See
//! `docs/specs/2026-09-15-turnstile-and-queues.md`.
//!
//! ## Why there is no `receive`/`consume` method
//!
//! Cloudflare Queues, like most managed queues fronting a serverless
//! compute model, deliver a consumer's messages by **invoking** it — a
//! Worker exports a `queue(batch, env)` handler that the platform calls with
//! a batch, rather than the consumer polling a `receive()` endpoint. A pull
//! API would misdescribe how delivery actually works and would not be
//! implementable against the one real vendor this contract has. Consuming a
//! queue is an entry point a product's own code exports, not a call an
//! adapter makes — the same shape of boundary `deploy` already draws around
//! `wrangler deploy`.

use thiserror::Error;

use crate::BoxFuture;

/// Errors a [`Queue`] implementation may return.
#[derive(Debug, Error)]
pub enum QueueError {
    #[error("send failed: {0}")]
    Send(String),
    #[error("{0}")]
    Other(String),
}

/// Asynchronous queue contract — producer side only.
pub trait Queue: Send + Sync {
    /// Enqueue a single message body.
    fn send<'a>(&'a self, body: &'a [u8]) -> BoxFuture<'a, Result<(), QueueError>>;

    /// Enqueue several message bodies as one batch.
    ///
    /// Vendors that have no native batch primitive may send sequentially;
    /// the contract makes no atomicity promise across the batch (unlike
    /// [`crate::Database::batch`], which does, because every vendor behind
    /// that trait has a real transaction or batch primitive to use).
    fn send_batch<'a>(&'a self, bodies: &'a [&'a [u8]]) -> BoxFuture<'a, Result<(), QueueError>>;
}

// ── None implementation ───────────────────────────────────────────────────────

/// No-op queue — every send succeeds silently and the message is discarded.
///
/// `queue = "none"` is a real, working selection: a product not yet doing
/// background work compiles, runs, and is testable without a live queue.
pub struct NoneQueue;

impl Queue for NoneQueue {
    fn send<'a>(&'a self, _body: &'a [u8]) -> BoxFuture<'a, Result<(), QueueError>> {
        Box::pin(std::future::ready(Ok(())))
    }

    fn send_batch<'a>(&'a self, _bodies: &'a [&'a [u8]]) -> BoxFuture<'a, Result<(), QueueError>> {
        Box::pin(std::future::ready(Ok(())))
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn q() -> Box<dyn Queue> {
        Box::new(NoneQueue)
    }

    #[tokio::test]
    async fn send_succeeds_silently() {
        q().send(b"hello").await.unwrap();
    }

    #[tokio::test]
    async fn send_batch_succeeds_silently() {
        q().send_batch(&[b"a".as_slice(), b"b".as_slice()])
            .await
            .unwrap();
    }

    #[test]
    fn none_queue_is_object_safe() {
        let _: Box<dyn Queue> = Box::new(NoneQueue);
    }
}
